File: python/examples/send_decode.py
"""Sends a transaction v1 transfer to a validator and decodes it back.

Recovers the config from both the raw wire bytes and the RPC's JSON projection.

Run with `just py-send-decode`.
"""

from __future__ import annotations

import base64

from solders.hash import Hash
from solders.keypair import Keypair
from solders.message import MessageV1, to_bytes_versioned

from txv1.feature import assert_v1_active
from txv1.rpc import RpcClient, RpcError, format_transaction_config
from txv1.send import build_v1_transfer, funded_keypair, send_and_confirm
from txv1.v1 import EXAMPLE_CONFIG, decode_transaction_version, format_config


def main() -> None:
    with RpcClient() as client:
        assert_v1_active(client)

        payer = funded_keypair(client, 1_000_000_000)
        recipient = Keypair().pubkey()
        blockhash = Hash.from_string(client.get_latest_blockhash())

        # The compute budget is a property of the message, so the instruction
        # list holds only the transfer.
        transaction = build_v1_transfer(payer, recipient, 10_000_000, blockhash, EXAMPLE_CONFIG)
        message = transaction.message
        assert isinstance(message, MessageV1)

        # `bytes()` on the message omits the version prefix; `to_bytes_versioned`
        # is what goes on the wire, and v1 announces itself with 0x81.
        wire_message = to_bytes_versioned(message)
        print("== compiled locally ==")
        print(f"  message bytes:      {len(wire_message)}")
        print(f"  version prefix:     0x{wire_message[0]:02x}")
        print(f"  instructions:       {len(message.instructions)}")
        print(f"  lifetime_specifier: {message.lifetime_specifier}")
        print(format_config(message.config))

        signature = send_and_confirm(client, transaction)
        print(f"\n== sent ==\n  signature: {signature}")

        # Without `maxSupportedTransactionVersion` the caller is capped at
        # legacy, and the server refuses outright rather than degrading.
        print("\n== getTransaction without maxSupportedTransactionVersion ==")
        try:
            client.call(
                "getTransaction", [signature, {"commitment": "confirmed", "encoding": "base64"}]
            )
            print("  unexpectedly succeeded")
        except RpcError as error:
            print(f"  rejected: {error}")

        encoded = client.call(
            "getTransaction",
            [
                signature,
                {
                    "commitment": "confirmed",
                    "encoding": "base64",
                    "maxSupportedTransactionVersion": 1,
                },
            ],
        )
        version, config = decode_transaction_version(base64.b64decode(encoded["transaction"][0]))
        print("\n== decoded from base64 wire bytes ==")
        print(f"  version:            {version.value}")
        print(format_config(config))

        if config != message.config:
            raise RuntimeError("config did not survive the round trip")
        print("  config round-tripped exactly")

        fetched = client.call(
            "getTransaction",
            [
                signature,
                {
                    "commitment": "confirmed",
                    "encoding": "json",
                    "maxSupportedTransactionVersion": 1,
                },
            ],
        )
        print("\n== the same config as the RPC's JSON projection ==")
        print(f"  version:            {fetched['version']}")
        print(format_transaction_config(fetched["transaction"]["message"].get("transactionConfig")))


if __name__ == "__main__":
    main()

File: ts/web3js-v3/src/send-decode.ts
// Run with `just w3v3-send-decode` while a validator is up (`just validator-start`).

import { Connection, Keypair, LAMPORTS_PER_SOL, MessageV1, SystemProgram, VersionedTransaction } from '@solana/web3.js';

import { fundedKeypair, json, RPC_URL, sendV1Transaction } from './lib/v1';

const connection = new Connection(RPC_URL, 'confirmed');
const payer = await fundedKeypair(connection, LAMPORTS_PER_SOL);
const recipient = await Keypair.generate();

const { signature, transaction } = await sendV1Transaction(
    connection,
    payer,
    [
        SystemProgram.transfer({
            fromPubkey: payer.publicKey,
            lamports: 10_000_000,
            toPubkey: recipient.publicKey,
        }),
    ],
    { computeUnitLimit: 20_000, heapSize: 64 * 1024, loadedAccountsDataSizeLimit: 64 * 1024 },
);

console.log('== sent ==');
console.log(`  signature: ${signature}`);
console.log(`  version:   ${transaction.version}`);
console.log(`  size:      ${transaction.serialize().length} bytes`);

console.log('\n== getTransaction with maxSupportedTransactionVersion 0 ==');
try {
    await connection.getTransaction(signature, { maxSupportedTransactionVersion: 0 });
    console.log('  unexpectedly succeeded');
} catch (error) {
    console.log(`  rejected: ${error instanceof Error ? error.message : String(error)}`);
}

const fetched = await connection.getTransaction(signature, { maxSupportedTransactionVersion: 1 });
if (fetched === null) {
    throw new Error(`transaction ${signature} was not found`);
}

const { message } = fetched.transaction;
console.log('\n== read back ==');
console.log(`  version: ${message.version} (${fetched.version} in the response envelope)`);
console.log(`  fee:     ${fetched.meta?.fee} lamports`);
if (!(message instanceof MessageV1)) {
    throw new Error('the transaction just sent did not come back as a v1 message');
}
console.log(`  config:  ${json(message.transactionConfig)}`);

const roundTripped = VersionedTransaction.deserialize(transaction.serialize());
console.log('\n== decoded from the local wire bytes ==');
console.log(`  version:      ${roundTripped.version}`);
console.log(`  instructions: ${roundTripped.message.compiledInstructions.length}`);
console.log(
    `  config:       ${roundTripped.message instanceof MessageV1 ? json(roundTripped.message.transactionConfig) : 'none'}`,
);
